optuna_dashboard/_app.py

```diff
@@ -411,6 +412,47 @@ def save_study_note(study_id: int) -> BottleViewReturn:
         response.status = 204  # No content
         return {}
 
+    @app.post("/api/studies/<study_id:int>/<trial_id:int>/tell")
+    @json_api_view
+    def tell_trial(study_id: int, trial_id: int) -> BottleViewReturn:
+        s = request.json.get("state", None)
+        vs = request.json.get("values", None)
+
+        try:
+            values = [float(v) for v in vs] if vs is not None else vs
+        except ValueError:
+            response.status = 400  # Bad request
+            return {"reason": "You need to pass float castable values"}
+
+        string2State = {
+            "Running": TrialState.RUNNING,
+            "Complete": TrialState.COMPLETE,
+            "Pruned": TrialState.PRUNED,
+            "Fail": TrialState.FAIL,
+            "Waiting": TrialState.WAITING,
+        }
+        if s not in string2State:
+            response.status = 400  # Bad request
+            return {"reason": f"You passed {s} as a state, which is not defined in Optuna."}
+        state = string2State[s]
+
+        try:
+            study_name = storage.get_study_name_from_id(study_id)
+        except KeyError:
+            response.status = 404  # Not found
+            return {"reason": f"study_id={study_id} is not found"}
+
+        study = optuna.load_study(storage=storage, study_name=study_name)
+
+        try:
+            study.tell(trial_id, values=values, state=state)
```

Comment:
Would you mind motivating this change a bit? 🙏

Reply:
Since the `Study` class is designed for Optuna users, some code may be unnecessary or inefficient in Optuna Dashboard.

For example, the following code is inefficient since it cannot use the trial cache that Optuna Dashboard has.
https://github.com/optuna/optuna/blob/master/optuna/study/_tell.py#L103

Also, the input values are validated in `study.tell()` function, but they need to be validated again in order for Optuna Dashboard's API server to return the proper response status and error messages.
https://github.com/optuna/optuna/blob/master/optuna/study/_tell.py#L127-L158

Reply:
To use `storage.get_trial_state_values`, it also requires some more lines. For example, https://github.com/optuna/optuna/blob/master/optuna/study/_tell.py#L168-L173 . In my opinion, it could be safer to rely on the `study.tell()` function because the optuna may add more code into the tell functionality.

Reply:
Regarding the lines you commented, optuna-dashboard does not require them. I'm not sure if `study.tell()` will contain the code that Optuna Dashboard would require in the future, but at this time there is no such code. Rather, it contains code that is clearly unnecessary or inefficient. I would say we should not rely on `study.tell()` since it is designed for Optuna users, not an Optuna Dashboard.

What do you think?

Reply:
Thanks for the clarification, and the discussion. 

Personally, it's fine to use the storage API directly, at least in this case since it's not a private method, and I understand that the dashboard calls "Study.tell" under different circumstances, for instance doesn't allow specifying the state. I'm slightly skeptical about having validation logic at several places, though, more than the speed gained by using the cache, since the study also checks for NaN, the number of objectives in the case of MO, etc. If we'd in the future allow other states such as pruning, then I'd suggest reconsidering using `Study.tell` here.

Reply:
Ok. I understand. I followed it.